docs/backend/config.md
---
id: config
title: Configuration
sidebar_label: Configuration
---

This is an ```object``` that needs to be given as a parameter to the ```init``` function when your node server starts.

This object has the following shape:
```ts
// "?" means that parameter is optional
let config = {
    mysql: {
        host?: string,  // default localhost
        port?: number, // default 3306
        user: string, // If the tables in the database are not created already, then this user must have permission to create tables.
        password: string,
        connectionLimit?: number, // default 50
        database: string, // name of the database to connect to. This must be created before running this package
        tables?: {
            signingKey?: string, // default signing_key - table name used to store secret keys
            refreshTokens?: string // default refresh_token - table name used to store sessions
        }
    },
    tokens: {
        accessToken?: {
            signingKey?: {
                dynamic?: boolean, // default true - if this is true, then the JWT signing key will change automatically ever updateInterval hours (Unless you use the get function mentioned below).
                updateInterval?: number, // in hours - default 24 - should be >= 1 && <= 720. Determines how often to change the signing key. If dynamic is false, then this does not matter. 
                get?: () => Promise<string>, // default undefined - If you want to give your own JWT signing key, please give a function here. If this is given, then the dynamic boolean will be ignored as key management will be up to you. This function will be called every time we generate or verify any JWT, so please make sure it is efficient.
            },
            validity?: number, // in seconds, default is 3600 seconds. Should be >= 10 && <= 86400000 seconds. This determines the lifetime of an access token.
            blacklisting?: boolean, // default is false. If you set this to true, revoking a session will cause immediate logout of the user using that session, regardless of access token's lifetime (Their access token will be invalidated). But know that this has an adverse effect on time efficiency of each getSession call.
            accessTokenPath?: string // default is "/". this will be the path of the access token cookie.
        },
        refreshToken: {
            validity?: number, // in hours, default is 2400 (100 days). Should be >= 1 hour && <= 365 * 24 hours. This determines how long a refresh token is alive for. So if your user is inactive for these many hours, they will be logged out.
            removalCronjobInterval?: string, // in the same style as of crontab, but with an extra seconds field as well. Default is "0 0 0 1-31/7 * *" - every 7th day of the month from 1 through 31. Defines how often the cronjob that removes expired sessions from the db should run.
            renewTokenPath: string // this is the API path that needs to be called for refreshing a session. This needs to be a POST API. An example value is "/api/refreshtoken". This will also be the path of the refresh token cookie.
        },
        enableAntiCsrf?: boolean // default is true. When set to true, an anti-csrf token will be sent in header during create/refresh session function calls. if set to false, anti-csrf token will not be generated and if any value passed in header as anti-csrf token, that will be ignored. we strongly recommend to set this value to true. 
    },
    logging?: {
        info?: (info: any) => void, // default undefined. This function, if provided, will be called for info logging purposes
        error?: (err: any) => void // default undefined. This function, if provided, will be called for error logging purposes
    },
    cookie: {
        domain: string, // this is the domain to set for all the cookies. For example, "supertokens.io" The path for all cookies except the refresh token will be "/"
        secure?: boolean // default true. Sets if the cookies are secure or not. Ideally, this value should be true in production mode.
    }
}
```

To see how to use this object, please refer to the [Init & Imports](usage-with-express/initialisation#example-code) section.

## Development vs Production
- If your development mode does not support ```https```, please be sure to set the ```secure``` option to ```false```. Otherwise your access and refresh tokens will not be sent.
- In production, we highly recommend that you use ```https``` everywhere. In fact, you should redirect all ```http``` requests to use ```https``` as well!

docs/backend-setup/config.md
---
id: config
hide_title: true
title: Configuration
sidebar_label: Configuration
---



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [4.1.0] - 2019-07-17
### Fixed
- Throws Auth error if userId is not of correct type
- Adds feature for options API.
- CSRF protection can be disabled (by default it will be enabled)
- userId can be either string or number

## [4.0.4] - 2019-07-15
### Fixed
- Uses accessTokenPath for idRefreshToken as well.

## [4.0.3] - 2019-07-15
### Fixed
- MySQL connection error will now show actual error that was thrown when trying to establish the connection.

## [4.0.2] - 2019-07-14
### Changed
- Adds support for when API is a different subdomain to website URL by using Access-Control-Allow-Credentials, Access-Control-Allow-Headers and Access-Control-Expose-Headers.

## [4.0.1] - 2019-07-12
### Changed
- CSRF protection can be disabled (by default it will be enabled)
- userId can be either string or number

## [4.0.0] - 2019-07-09
### Breaking Changes
- Implements CSRF protection
- Makes handing token theft via an error as opposed to a callback

## [3.0.0] - 2019-06-15
### Breaking Changes
- Reorganises lib so that express related functions are access via /express. By default, you get the raw functions so that you can use them with any non express frameworks as well.
- Stores sessionHandle without hashing in db - results in changing of column name in table

## [2.0.3] - 2019-06-15
### Added
- Does not store hashed version of session handle in database anymore.

## [2.0.2] - 2019-06-14
### Changed
- Adds feature of blacklisting, and if JWT verification fails, then reloads key from DB.

## [2.0.1] - 2019-06-12
### Changed
- Replaces usage of cron with node-cron - to reduce the overall size of this package

## [2.0.0] - 2019-06-07
### Changed
- Changed function signatures in session.ts to remove ```idRefreshToken``` if the function body does not require it. The version has been changed to 2.x.x because indexRaw.ts exposes the functions in session.ts - so any service using that, would break too if version was still 1.x.x

## [1.0.14] - 2019-06-4
### Changed
- Moved the cronjob timing param to the input config object

## [1.0.12] - 2019-05-31
### Changed
- Update readme.


lib/ts/config.ts
import { AuthError, generateError } from "./error";
import { TypeConfig, TypeGetSigningKeyUserFunction, TypeInputConfig } from "./helpers/types";
import { sanitizeBooleanInput, sanitizeNumberInput, sanitizeStringInput } from "./helpers/utils";

/**
 * @class Config
 * @description this is a singleton class since we need just one Config for this node process.
 */
export default class Config {
    private static instance: undefined | Config;
    private config: TypeConfig;

    private constructor(config: TypeConfig) {
        this.config = config;
    }

    /**
     * @description called when library is being initialised
     * @throws AuthError GENERAL_ERROR
     */
    static init(config: TypeInputConfig) {
        if (Config.instance === undefined) {
            Config.instance = new Config(setDefaults(validateAndNormalise(config)));
        }
    }

    static get(): TypeConfig {
        if (Config.instance === undefined) {
            throw generateError(
                AuthError.GENERAL_ERROR,
                new Error("configs not set. Please call the init function before using this library"),
                false
            );
        }
        return Config.instance.config;
    }

    static reset = () => {
        if (process.env.TEST_MODE !== "testing") {
            throw Error("call this function only during testing");
        }
        Config.instance = undefined;
    };

    static isInitialised = () => {
        if (process.env.TEST_MODE !== "testing") {
            throw Error("call this function only during testing");
        }
        return Config.instance !== undefined;
    };
}

/**
 * @description checks for user input validity in terms of types and ranges
 * @throws AuthError GENERAL_ERROR
 */
const validateAndNormalise = (config: any): TypeInputConfig => {
    if (config === null || typeof config !== "object") {
        throw generateError(AuthError.GENERAL_ERROR, new Error("passed config is not an object"), false);
    }
    const mysqlInputConfig = config.mysql;
    if (typeof mysqlInputConfig !== "object") {
        throw generateError(
            AuthError.GENERAL_ERROR,
            new Error("mysql config not passed. user, password and database are required"),
            false
        );
    }
    const host = sanitizeStringInput(mysqlInputConfig.host);
    const port = sanitizeNumberInput(mysqlInputConfig.port);
    const user = sanitizeStringInput(mysqlInputConfig.user);
    if (user === undefined) {
        throw generateError(AuthError.GENERAL_ERROR, new Error("mysql config error. user not passed"), false);
    }
    const password = sanitizeStringInput(mysqlInputConfig.password);
    if (password === undefined) {
        throw generateError(AuthError.GENERAL_ERROR, new Error("mysql config error. password not passed"), false);
    }
    const connectionLimit = sanitizeNumberInput(mysqlInputConfig.connectionLimit);
    const database = sanitizeStringInput(mysqlInputConfig.database);
    if (database === undefined) {
        throw generateError(AuthError.GENERAL_ERROR, new Error("mysql config error. database not passed"), false);
    }
    let tables:
        | {
              signingKey: string | undefined;
              refreshTokens: string | undefined;
          }
        | undefined;
    const tablesMysqlInputConfig = mysqlInputConfig.tables;
    if (tablesMysqlInputConfig !== undefined) {
        const signingKey = sanitizeStringInput(tablesMysqlInputConfig.signingKey);
        const refreshTokens = sanitizeStringInput(tablesMysqlInputConfig.refreshTokens);
        tables = {
            signingKey,
            refreshTokens
        };
    }
    const mysql = {
        host,
        port,
        user,
        password,
        connectionLimit,
        database,
        tables
    };
    let tokensInputConfig = config.tokens;
    const accessTokenInputConfig = tokensInputConfig.accessToken;
    let accessToken:
        | {
              signingKey:
                  | {
                        dynamic: boolean | undefined;
                        updateInterval: number | undefined;
                        get: TypeGetSigningKeyUserFunction | undefined;
                    }
                  | undefined;
              validity: number | undefined;
              blacklisting: boolean | undefined;
              accessTokenPath: string | undefined;
              antiCsrf: boolean | undefined;
          }
        | undefined;
    if (accessTokenInputConfig !== undefined) {
        const signingKeyInputConfig = accessTokenInputConfig.signingKey;
        let signingKey:
            | {
                  dynamic: boolean | undefined;
                  updateInterval: number | undefined;
                  get: TypeGetSigningKeyUserFunction | undefined;
              }
            | undefined;
        if (signingKeyInputConfig !== undefined) {
            const dynamic = sanitizeBooleanInput(signingKeyInputConfig.dynamic);
            let updateInterval = sanitizeNumberInput(signingKeyInputConfig.updateInterval);
            if (updateInterval !== undefined && process.env.TEST_MODE !== "testing") {
                if (updateInterval > defaultConfig.tokens.accessToken.signingKey.updateInterval.max) {
                    throw generateError(
                        AuthError.GENERAL_ERROR,
                        new Error(
                            "update interval passed for updating singingKey for access token is not within allowed interval. (Note: value passed will be in units of hours)"
                        ),
                        false
                    );
                } else if (updateInterval < defaultConfig.tokens.accessToken.signingKey.updateInterval.min) {
                    throw generateError(
                        AuthError.GENERAL_ERROR,
                        new Error(
                            "update interval passed for updating singingKey for access token is not within allowed interval. (Note: value passed will be in units of hours)"
                        ),
                        false
                    );
                }
            }
            const get = signingKeyInputConfig.get;
            if (get !== undefined && typeof get !== "function") {
                throw generateError(
                    AuthError.GENERAL_ERROR,
                    new Error("config > tokens > accessToken > get must be a function"),
                    false
                );
            }
            signingKey = {
                dynamic,
                updateInterval,
                get
            };
        }
        let validity = sanitizeNumberInput(accessTokenInputConfig.validity);
        if (validity !== undefined && process.env.TEST_MODE !== "testing") {
            if (validity > defaultConfig.tokens.accessToken.validity.max) {
                throw generateError(
                    AuthError.GENERAL_ERROR,
                    new Error(
                        "passed value for validity of access token is not within allowed interval. (Note: value passed will be in units of seconds)"
                    ),
                    false
                );
            } else if (validity < defaultConfig.tokens.accessToken.validity.min) {
                throw generateError(
                    AuthError.GENERAL_ERROR,
                    new Error(
                        "passed value for validity of access token is not within allowed interval. (Note: value passed will be in units of seconds)"
                    ),
                    false
                );
            }
        }
        let blacklisting = sanitizeBooleanInput(accessTokenInputConfig.blacklisting);
        let accessTokenPath = sanitizeStringInput(accessTokenInputConfig.accessTokenPath);
        let antiCsrf = sanitizeBooleanInput(accessTokenInputConfig.antiCsrf);
        accessToken = {
            signingKey,
            validity,
            blacklisting,
            accessTokenPath,
            antiCsrf
        };
    }
    let refreshTokenInputConfig = tokensInputConfig.refreshToken;
    if (typeof refreshTokenInputConfig !== "object") {
        throw generateError(
            AuthError.GENERAL_ERROR,
            new Error("refreshToken config not passed. renewTokenPath is required"),
            false
        );
    }
    let validity = sanitizeNumberInput(refreshTokenInputConfig.validity);
    if (validity !== undefined && process.env.TEST_MODE !== "testing") {
        if (validity > defaultConfig.tokens.refreshToken.validity.max) {
            throw generateError(
                AuthError.GENERAL_ERROR,
                new Error(
                    "passed value for validity of refresh token is not within allowed interval. (Note: value passed will be in units of hours)"
                ),
                false
            );
        } else if (validity < defaultConfig.tokens.refreshToken.validity.min) {
            throw generateError(
                AuthError.GENERAL_ERROR,
                new Error(
                    "passed value for validity of refresh token is not within allowed interval. (Note: value passed will be in units of hours)"
                ),
                false
            );
        }
    }
    const renewTokenPath = sanitizeStringInput(refreshTokenInputConfig.renewTokenPath);
    if (renewTokenPath === undefined) {
        throw generateError(AuthError.GENERAL_ERROR, new Error("renewTokenPath not passed"), false);
    }
    const removalCronjobInterval = sanitizeStringInput(refreshTokenInputConfig.removalCronjobInterval);
    if (removalCronjobInterval !== undefined && typeof removalCronjobInterval !== "string") {
        throw generateError(
            AuthError.GENERAL_ERROR,
            new Error("removalCronjobInterval does not have the correct type"),
            false
        );
    }
    const refreshToken = {
        renewTokenPath,
        removalCronjobInterval,
        validity
    };
    const tokens = {
        accessToken,
        refreshToken
    };
    let loggingInputConfig = config.logging;
    let logging;
    if (loggingInputConfig !== undefined) {
        let info = loggingInputConfig.info;
        let error = loggingInputConfig.error;
        if (info !== undefined && typeof info !== "function") {
            throw generateError(
                AuthError.GENERAL_ERROR,
                new Error("logging config error. info option passed must be a function"),
                false
            );
        }
        if (error !== undefined && typeof error !== "function") {
            throw generateError(
                AuthError.GENERAL_ERROR,
                new Error("logging config error. error option passed must be a function"),
                false
            );
        }
        logging = {
            info,
            error
        };
    }
    const cookieInputConfig = config.cookie;
    const domain = sanitizeStringInput(cookieInputConfig.domain);
    const secure = sanitizeBooleanInput(cookieInputConfig.secure);
    if (domain === undefined) {
        throw generateError(AuthError.GENERAL_ERROR, new Error("domain parameter for cookie not passed"), false);
    }
    const cookie = {
        domain,
        secure
    };
    return {
        mysql,
        tokens,
        cookie,
        logging
    };
};

const setDefaults = (config: TypeInputConfig): TypeConfig => {
    // TODO: change this style of a || b to a === undefined ? b : a
    return {
        mysql: {
            host: config.mysql.host || defaultConfig.mysql.host,
            port: config.mysql.port || defaultConfig.mysql.port,
            user: config.mysql.user,
            password: config.mysql.password,
            connectionLimit: config.mysql.connectionLimit || defaultConfig.mysql.connectionLimit,
            database: config.mysql.database,
            tables:
                config.mysql.tables === undefined
                    ? defaultConfig.mysql.tables
                    : {
                          refreshTokens: config.mysql.tables.refreshTokens || defaultConfig.mysql.tables.refreshTokens,
                          signingKey: config.mysql.tables.signingKey || defaultConfig.mysql.tables.signingKey
                      }
        },
        tokens: {
            accessToken:
                config.tokens.accessToken === undefined
                    ? {
                          signingKey: {
                              dynamic: defaultConfig.tokens.accessToken.signingKey.dynamic,
                              updateInterval:
                                  defaultConfig.tokens.accessToken.signingKey.updateInterval.default * 60 * 60 * 1000,
                              get: undefined
                          },
                          validity: defaultConfig.tokens.accessToken.validity.default * 1000,
                          blacklisting: defaultConfig.tokens.accessToken.blacklisting,
                          accessTokenPath: defaultConfig.tokens.accessToken.accessTokenPath,
                          antiCsrf: defaultConfig.tokens.accessToken.antiCsrf
                      }
                    : {
                          signingKey:
                              config.tokens.accessToken.signingKey === undefined
                                  ? {
                                        dynamic: defaultConfig.tokens.accessToken.signingKey.dynamic,
                                        updateInterval:
                                            defaultConfig.tokens.accessToken.signingKey.updateInterval.default *
                                            60 *
                                            60 *
                                            1000,
                                        get: undefined
                                    }
                                  : {
                                        dynamic:
                                            config.tokens.accessToken.signingKey.dynamic === undefined
                                                ? defaultConfig.tokens.accessToken.signingKey.dynamic
                                                : config.tokens.accessToken.signingKey.dynamic,
                                        updateInterval:
                                            (config.tokens.accessToken.signingKey.updateInterval ||
                                                defaultConfig.tokens.accessToken.signingKey.updateInterval.default) *
                                            60 *
                                            60 *
                                            1000,
                                        get: config.tokens.accessToken.signingKey.get
                                    },
                          validity:
                              (config.tokens.accessToken.validity ||
                                  defaultConfig.tokens.accessToken.validity.default) * 1000,
                          blacklisting:
                              config.tokens.accessToken.blacklisting === undefined
                                  ? defaultConfig.tokens.accessToken.blacklisting
                                  : config.tokens.accessToken.blacklisting,
                          accessTokenPath:
                              config.tokens.accessToken.accessTokenPath === undefined
                                  ? defaultConfig.tokens.accessToken.accessTokenPath
                                  : config.tokens.accessToken.accessTokenPath,
                          antiCsrf:
                              config.tokens.accessToken.antiCsrf === undefined
                                  ? defaultConfig.tokens.accessToken.antiCsrf
                                  : config.tokens.accessToken.antiCsrf
                      },
            refreshToken: {
                validity:
                    (config.tokens.refreshToken.validity || defaultConfig.tokens.refreshToken.validity.default) *
                    60 *
                    60 *
                    1000,
                removalCronjobInterval:
                    config.tokens.refreshToken.removalCronjobInterval === undefined
                        ? defaultConfig.tokens.refreshToken.removalCronjobInterval
                        : config.tokens.refreshToken.removalCronjobInterval,
                renewTokenPath: config.tokens.refreshToken.renewTokenPath
            }
        },
        cookie: {
            secure: config.cookie.secure === undefined ? defaultConfig.cookie.secure : config.cookie.secure,
            domain: config.cookie.domain
        },
        logging: {
            info: config.logging !== undefined ? config.logging.info : undefined,
            error: config.logging !== undefined ? config.logging.error : undefined
        }
    };
};

const defaultConfig = {
    mysql: {
        host: "localhost",
        port: 3306,
        connectionLimit: 50,
        tables: {
            signingKey: "signing_key",
            refreshTokens: "refresh_token"
        }
    },
    tokens: {
        accessToken: {
            signingKey: {
                dynamic: true,
                updateInterval: {
                    // in hours.
                    min: 1,
                    max: 720,
                    default: 24
                }
            },
            validity: {
                // in seconds
                min: 10,
                max: 1000 * 24 * 3600,
                default: 3600
            },
            blacklisting: false,
            accessTokenPath: "/",
            antiCsrf: true
        },
        refreshToken: {
            validity: {
                // in hours.
                min: 1,
                max: 365 * 24,
                default: 100 * 24
            },
            removalCronjobInterval: "0 0 0 1-31/7 * *" // every 7th day of a month starting from the 1st until the 31st
        }
    },
    cookie: {
        secure: true
    }
};
